Add porting guide and documentation for changes to argument spec validation

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Add detailed information about breaking changes to the 2.11 porting guide. Update the `module_utils` doc to include documentation for a lot more of the public interfaces.

Fixes #73983.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Docs Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`docs/docsite/rst/porting_guides/porting_guide_core_2.11.rst`
`docs/docsite/rst/reference_appendices/module_utils.rst`
`lib/ansible/module_utils/common/arg_spec.py`
`lib/ansible/module_utils/common/parameters.py`

samdoran committed:
Add ArgumentSpecValidator to docs

samdoran committed:
Improve docs for ArgumentSpecValidator

samdoran committed:
Document removal of private methods

samdoran committed:
Update module_utils documentation

Add docs for argument spec classes as well as validation and parameters files.

samdoran committed:
Update docs in arg_spec and paremeters

This improves the generated documentation.

samdoran committed:
Document breaking changes in porting guide.

samdoran committed:
Update formatting in porting guide and add a Deprecated section

samdoran committed:
Can't reference methods/functions that don't exist

Just use inline code style instead.

samdoran committed:
Fine tune module_utils documentation

- preserve the order in the source for errors.py
- document DEFAULT_TYPE_VALIDATORS so it can be referenced elsewhere
- fix automodule directive for validation.py

samdoran committed:
Improve arg_spec and validation doc strings

samdoran commented:
Our docs sanity test does not like specifying optional parameters.

```rst

:param: some_param
:type some_param: list, optional
```

I'm pretty sure that is valid in Sphinx.


```
/root/ansible/lib/ansible/module_utils/common/arg_spec.py:docstring of ansible.module_utils.common.arg_spec.ArgumentSpecValidator:: WARNING: py:class reference target not found: optional
```

abadger commented:
> Our docs sanity test does not like specifying optional parameters.
> 
> ```
> :param: some_param
> :type some_param: list, optional
> ```
> 
> I'm pretty sure that is valid in Sphinx.
> 
> ```
> /root/ansible/lib/ansible/module_utils/common/arg_spec.py:docstring of ansible.module_utils.common.arg_spec.ArgumentSpecValidator:: WARNING: py:class reference target not found: optional
> ```
> 
> It renders correctly, but the sanity test doesn't like it.

As a workaround, you can use:
``` rst
:arg positional_manadatory_arg: Description of argument
:type: str
:kwarg keyword_optional_arg: Description of argument
:type list:
```

I've never used the `:param:` syntax before so I don't know what cornercase this might be running into.

abadger commented:
Another thing to try would be to document the class construction parameters in the `__init__()` method's docstring rather than in the class docstring.  It might be that using `:type: foo, bar` in the `:py:class:` directive isn't working right but it would work right in the `:py:meth:` and `:py:func:` directives.

samdoran commented:
> I've never used the :param: syntax before so I don't know what cornercase this might be running into.

My understanding is that `arg`, `kwarg`, and `param` all mean the same thing and render the same in the output. That's what I gather from [the field lists documentation](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists). My testing shows that they all get grouped under **Parameters** in the rendered doc.

> Another thing to try would be to document the class construction parameters in the __init__() method's docstring rather than in the class docstring. 

That didn't change anything.

I think the root of this is that specifying `optional` is incorrect. I found that example [here](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html#the-sphinx-docstring-format) and maybe it is wrong.

samdoran committed:
Move instance docstring to the __init__ method

Remove optional description since it fails the sanity test and I am not 100% it is
valid anyway.

samdoran committed:
Remoe incorrect parameter from docstring

This was changed a while ago but wasn't removed from the docstring.

samdoran committed:
Use attr rather than attribute

The py:attribute: domain only exists in newer Sphinx >= 3.1.

samdoran committed:
Improve documentation on exceptions

webknjaz commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -114,6 +106,31 @@ def __init__(self, argument_spec,
+
+        :param mutually_exclusive: List or list of lists of terms that should not
+            be provided together.
+        :type mutually_exclusive: list
```
Try
```suggestion
        :type mutually_exclusive: Optional[list]
```

samdoran replied:
Wouldn't this require having imported `Optional` from `typing`? Using this fails the docs-build sanity test.

webknjaz commented:
@samdoran @abadger Sphinx build runs on py3-only already right? Can we add https://pypi.org/project/sphinx-autodoc-typehints/ and use type hints instead of putting types in docstrings?

webknjaz commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -130,30 +147,37 @@ def __init__(self, argument_spec,
-        Error messages in the ValidationResult may contain no_log values and should be
-        sanitized before logging or displaying.
+        Error messages in the :class:`ValidationResult` may contain no_log values and should be
+        sanitized with :func:`ansible.module_utils.common.parameters.sanitize_keys` before logging or displaying.
```
maybe render it shorter?
```suggestion
        sanitized with :func:`~ansible.module_utils.common.parameters.sanitize_keys` before logging or displaying.
```

webknjaz commented on lib/ansible/module_utils/errors.py:
```diff
@@ -39,16 +42,20 @@ def __delitem__(self, key):
         return [err.msg for err in self.errors]
 
     def append(self, error):
+        """Add a new error to ``self.errors``. Only :class:`AnsibleValidationError` should be added."""
```
pls only put a title on the first line per PEP257

webknjaz commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -114,6 +106,31 @@ def __init__(self, argument_spec,
+
+        :param required_by: Dictionary of parameter names that contain a list of
+            parameters required by each key in the dictionary.
+        :type required_by: dict
```
Sounds like it's
```suggestion
        :type required_by: dict[str, list[str]]
```

webknjaz commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -114,6 +106,31 @@ def __init__(self, argument_spec,
+
+        :param required_one_of: List of lists of terms, one of which in each list
+            is required.
+        :type required_one_of: list
```
is this
```suggestion
        :type required_one_of: list[list[str]]
```

webknjaz commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -114,6 +106,31 @@ def __init__(self, argument_spec,
+        :type mutually_exclusive: list
+
+        :kwarg required_together: List of lists of terms that are required together.
+        :type required_together: list
```
Or maybe
```suggestion
        :type required_together: list or None
```
per https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#python-signatures

abadger commented:
> > I've never used the :param: syntax before so I don't know what cornercase this might be running into.
> 
> My understanding is that `arg`, `kwarg`, and `param` all mean the same thing and render the same in the output. That's what I gather from [the field lists documentation](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists). My testing shows that they all get grouped under **Parameters** in the rendered doc.
> 
<nod> The reason I suggested using kwarg is that in that style, you don't have to say `, optional`.  I suppose you don't have to change from param as the keyword and can just leave optional off. (the rendered result is the same.  The wording "Default value is" in the description and the definition which shows there is a default value will both show that it is optional.

abadger commented:
Hmmmm.... One other thing that I thing isn't clear and is being hidden in the discussion of using type hints....

`optional` is not the same as `typing.Optional`.

When I read:
```
:param mutually_exclusive: Set to a list of pairs of parameters which are mutually exclusive with each other
:type: sequence of tuples, optional
```
I take it to mean that mutually_exclusive doesn't need to be specified.

`t.Optional` actually means what @webknjaz said `list of tuples or None`.

Delving into the implementation a bit (even though, that isn't what this PR is for ;-)   I think it would have been better to default to an empty tuple rather than None.  The two reasons I think that's better API is that (1) you then only accept sequences here, not sequences or None.  This is a bit less for API consumers to remember. and (2) It simplifies the logic of the API implementors.  If it's always a sequence, your code to process it looks like this:
``` python
for pairs in mutually_exclusive:
    # do thing
```

If you use None as the default value, you instead have to code like this:
``` python
if mutually_exclusive:
    for pairs in mutually_exclusive:
        # do thing
```

samdoran commented:
> I take it to mean that mutually_exclusive doesn't need to be specified.

Yup, that's what I was trying to convey. At this point, I'm just going to leave it to the function signature to convey that since we don't currently have a clear way to express it.

> I think it would have been better to default to an empty tuple rather than None.

That's a very good point! I tend to default to `None` or a string to avoid the [mutable defaults problem](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments)). I definitely should have used a tuple. 🤦‍♂️

samdoran committed:
Final pass

- use args/kwargs instead of param
- fix formatting errors that didn't display examples correctly
- format TypeErrors so they are referenced as classes
- specify complex types

samdoran committed:
Fix inline markup.

webknjaz commented:
@abadger yep, you are right. I actually often just say that it's optional in text. Also, the defaults are normally visible in the function signature line and it is pretty obvious what that means...

webknjaz commented:
Interestingly https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists has an example with `priority=1` in the signature and `int or Note` in the docstring 🤯

webknjaz commented on lib/ansible/module_utils/common/arg_spec.py:
```diff
@@ -114,6 +107,31 @@ def __init__(self, argument_spec,
+
+        :kwarg mutually_exclusive: List or list of lists of terms that should not
+            be provided together.
+        :type mutually_exclusive: list or list[list]
```
it's probably this?
```suggestion
        :type mutually_exclusive: list[str] or list[list[str]]
```

webknjaz commented on docs/docsite/rst/reference_appendices/module_utils.rst:
```diff
@@ -25,3 +25,50 @@ To use this functionality, include ``import ansible.module_utils.basic`` in your
+ArgumentSpecValidator
+=====================
+
+.. autoclass:: ansible.module_utils.common.arg_spec.ArgumentSpecValidator
```
Side note: It's possible to add `.. currentmodule:: ansible.module_utils` at the top and avoid prefixing it in every definition.

webknjaz commented:
@samdoran if you'll go the type hints way, make sure you include the imports. Example: https://github.com/ansible/ansible/blob/3740d7b/lib/ansible/galaxy/dependency_resolution/providers.py#L11-L21.

samccann commented on docs/docsite/rst/reference_appendices/module_utils.rst:
```diff
@@ -25,3 +25,50 @@ To use this functionality, include ``import ansible.module_utils.basic`` in your
+.. autoclass:: ansible.module_utils.common.arg_spec.ValidationResult
+   :members:
+   :member-order: bysource
+   :private-members: _no_log_values  # This only works in sphinx >= 3.2. Otherwise it shows all private members with doc strings.
```
Since we aren't at 3.2 or above, is this going to cause display problems for 2.11 docs?

samdoran replied:
No, it will display just fine with Sphinx 2.1.2, but only because no other private things have docstrings.

This comment is mainly to document the behavior that took me quite some time to figure out what was going on. `:private-members:` was rendering the docs for _all_ private methods/attributes even if they weren't in the list. It's because it wasn't until Sphinx 3.2 that `:private-members:` honored the values passed in.

samccann commented:
Also, there seems to be a lot here that impacts collection maintainers. Consider adding an entry into the next Ansible Bullhorn if you haven't already.

samccann reviewed (approved):
docs portion LGTM

samdoran commented:
> Also, there seems to be a lot here that impacts collection maintainers. Consider adding an entry into the next Ansible Bullhorn if you haven't already.

I opened issues on every collection that was using the private methods mentioned in the porting guide.

samdoran committed:
Update type in docstring
